src/main/scala/org/spartanz/parserz/ParsersModule.scala

```diff
@@ -67,12 +92,27 @@ trait ParsersModule {
     final def fail[E, A](e: E): Grammar[Any, Nothing, E, A] =
       unit.mapPartial(e)(PartialFunction.empty, PartialFunction.empty)
 
+    final def fail[SI, SO, E, A](es: SI => (SO, E)): Grammar[SI, SO, E, A] =
+      unit.mapPartial(es)(PartialFunction.empty, PartialFunction.empty)
```

Comment:
Maybe `failStatefully`, so you can have straight-up `fail` that doesn't change state.